core/controlplane/config/templates/cloud-config-etcd

```diff
@@ -130,34 +134,202 @@ ssh_authorized_keys:
+
+      # Skip the `while` block below when the EBS volume is already attached to this EC2 instance
+      describe_volume_result=$(aws ec2 describe-volumes --region {{.Region}} --filters Name=tag:aws:cloudformation:stack-name,Values=$stack_name Name=tag:kube-aws:owner:role,Values=etcd Name=attachment.instance-id,Values=$instance_id --output json)
+      assumed_hostname=$(echo "$describe_volume_result" | jq -r ".Volumes[].Tags[] | select(.Key == \"kube-aws:etcd:hostname\").Value")
+
+      # Decide which volume to attach hence hostname to assume
+      while [ "$assumed_hostname" = "" ]; do
+        sleep 3
+
+        aws ec2 describe-volumes \
+          --region {{.Region}} \
+          --filters Name=tag:aws:cloudformation:stack-name,Values=$stack_name Name=tag:kube-aws:owner:role,Values=etcd Name=status,Values=available \
+          --output json \
+          > /var/run/coreos/available-volumes.json
+
+        cat /var/run/coreos/available-volumes.json | \
+          jq -r ".Volumes | map({ VolumeId : .VolumeId, AvailabilityZone: .AvailabilityZone, State: .State, Tags: .Tags, Name: (.Tags[] | select(.Key == \"kube-aws:etcd:hostname\") | .Value) }) | group_by(.Name) | select(all(.[].State == \"available\")) | [.[][]]" \
+          > /var/run/coreos/available-etcd-volumes.json
+
+        vol=$(cat /var/run/coreos/available-etcd-volumes.json | jq -r "[.[] | select(.AvailabilityZone == \"$az\")][0]")
+        if [ "$vol" = ""]; then
+          echo "no etcd volume available in availability zone $az"
+        fi
+
+        vol_id=$(echo "$vol" | jq -r ".VolumeId")
+        # See http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/device_naming.html for device naming
+        if aws ec2 attach-volume --region {{.Region}} --volume-id $vol_id --instance-id $instance_id --device "/dev/xvdf"; then
+          eip_alloc_id=$(echo "$vol" | jq -r ".Tags[] | select(.Key == \"kube-aws:etcd:eip-alloc-id\").Value")
+
+          # Assume the hostname corresponds to the data(=the EBS volume identified by the vol_id above)
+          assumed_hostname=$(echo "$vol" | jq -r ".Tags[] | select(.Key == \"kube-aws:etcd:hostname\").Value")
+        fi
+      done
+
+      if [ "$vol" = "" ]; then
+        vol=$(echo "$describe_volume_result" | jq -r ".Volumes[0]")
+      fi
+      vol_id=$(echo "$vol" | jq -r ".VolumeId")
+      eip_alloc_id=$(echo "$vol" | jq -r ".Tags[] | select(.Key == \"kube-aws:etcd:eip-alloc-id\").Value")
+
+      # Wait until the volume attachment completes
+      until [ "$volume_status" = ok ]; do
+        sleep 3
+        describe_volume_status_result=$(aws ec2 describe-volume-status --region {{.Region}} --volume-id $vol_id --output json)
+        volume_status=$(echo "$describe_volume_status_result" | jq -r ".VolumeStatuses[0].VolumeStatus.Status")
+      done
+
+      echo $assumed_hostname > /var/run/coreos/assumed_hostname
+      echo $eip_alloc_id > /var/run/coreos/eip_alloc_id
+
+  - path: /opt/bin/assume-etcd-hostname-with-private-ip
```

Comment:
if you go with ENI, then IP addresses will be fixed and Route53 records can be created once and for all by CF